neuraxle/pipeline.py

```diff
@@ -486,3 +490,178 @@ def should_resume(self, data_container: DataContainer) -> bool:
+        """
+        data_inputs = sub_pipeline.join_transform(data_container.data_inputs)
+        data_container.set_data_inputs(data_inputs)
+
+        current_ids = self.hash(data_container.current_ids, self.hyperparams, data_inputs)
+        data_container.set_current_ids(current_ids)
+
+        return data_container
+
+    def _handle_fit_transform_sub_pipeline(self, sub_pipeline, data_container) -> \
+            Tuple['MiniBatchSequentialPipeline', DataContainer]:
+        """
+        Fit Transform sub pipeline using join fit transform.
+
+        :param sub_pipeline: sub pipeline to be used to transform data container
+        :param data_container: data container to fit transform
+        :return: fitted self, transformed data container
+        """
+        _, data_inputs = sub_pipeline.join_fit_transform(
+            data_inputs=data_container.data_inputs,
+            expected_outputs=data_container.expected_outputs
+        )
+        data_container.set_data_inputs(data_inputs)
+
+        current_ids = self.hash(data_container.current_ids, self.hyperparams, data_inputs)
+        data_container.set_current_ids(current_ids)
+
+        return self, data_container
+
+    def _create_sub_pipelines(self) -> List['MiniBatchSequentialPipeline']:
+        """
+        Create sub pipelines by splitting the steps by the join type name.
+
+        :return: list of sub pipelines
+        """
+        sub_pipelines: List[MiniBatchSequentialPipeline] = self.split(BARRIER_STEP_NAME)
+        for sub_pipeline in sub_pipelines:
+            if not sub_pipeline.ends_with_type_name(BARRIER_STEP_NAME):
+                raise Exception(
+                    'At least one Barrier step needs to be at the end of a streaming pipeline. '.format(
+                        self.join_type_name)
+                )
+
+        return sub_pipelines
+
+    def join_transform(self, data_inputs: Iterable) -> Iterable:
+        """
+        Concatenate the transform output of each batch of self.batch_size together.
+
+        :param data_inputs:
+        :return:
+        """
+        outputs = []
+        for batch in convolved_1d(
+                iterable=data_inputs,
+                kernel_size=self.batch_size
+        ):
+            batch_outputs = super().transform(batch)
+            outputs.extend(batch_outputs)  # TODO: use a joiner here
+
+        return outputs
+
+    def join_fit_transform(self, data_inputs: Iterable) -> Tuple['MiniBatchSequentialPipeline', Iterable]:
+        """
+        Concatenate the fit transform output of each batch of self.batch_size together.
+
+        :param data_inputs:
+        :return: fitted self, transformed data inputs
+        """
+        outputs = []
+        for batch in convolved_1d(
+                iterable=data_inputs,
+                kernel_size=self.batch_size
+        ):
+            _, batch_outputs = super().fit_transform(batch)
+            outputs.extend(batch_outputs)  # TODO: use a joiner here
```

Comment:
use a joiner here to join the outputs together ? would you say this should be required or not ? @guillaume-chevalier

Reply:
@alexbrillant I just found what's missing in the barriers! 

Barriers indeed contain logic. They don't just simply throw exceptions as I first thought. 

The logic they are responsible for is how to wait for data and how to merge data. We need to develop that idea more. 

E.g.: some kind of barriers could be a queue as a solution to the [producer-consumer problem](https://en.wikipedia.org/wiki/Producer–consumer_problem). Please read on the producer-consumer problem. This means we could have streaming pipelines that are semi-parallelized and for which the step to prioritize is one that has something to consume and for which the queue barrier still has more empty room to accept the data container that the next step will consume. 

The OOP design for those kind of barriers is very hard if we want to support checkpoints and also if we want the logic to not be all inside the streaming pipeline class: barriers must know how to handle the things next to them, and they must know how to interrupt or not the execution flow of the streaming pipelines.

Reply:
Not that instead of doing `outputs.extend(batch_outputs)  # TODO: use a joiner here`, we could do `barrier.join(batch_outputs)` or something like that. The barrier would call something like `data_container.extend(batch_outputs_data_container)` from within perhaps.

Reply:
The OOP and design patterns problem here are so hard that there are papers on that lol: 
http://www.romulosilvadeoliveira.eng.br/artigos/Rodrigo-Romulo-Montez-Sccc2005.pdf
:rofl:

Reply:
Okey, that is what I thought it was good : "Not that instead of doing outputs.extend(batch_outputs) # TODO: use a joiner here, we could do barrier.join(batch_outputs) or something like that. The barrier would call something like data_container.extend(batch_outputs_data_container) from within perhaps."

Reply:
yeah lol that is not trivial at all @guillaume-chevalier